src/coreclr/jit/redundantbranchopts.cpp

```diff
@@ -691,6 +692,294 @@ bool Compiler::optJumpThread(BasicBlock* const block, BasicBlock* const domBlock
+//
+// Returns:
+//    true, if changes were made.
+//
+bool Compiler::optRedundantRelop(BasicBlock* const block)
+{
+    Statement* const stmt = block->lastStmt();
+
+    if (stmt == nullptr)
+    {
+        return false;
+    }
+
+    GenTree* const jumpTree = stmt->GetRootNode();
+
+    if (!jumpTree->OperIs(GT_JTRUE))
+    {
+        return false;
+    }
+
+    GenTree* const tree = jumpTree->AsOp()->gtOp1;
+
+    if (!tree->OperIsCompare())
+    {
+        return false;
+    }
+
+    // If tree has side effects other than GTF_EXCEPT, bail.
+    //
+    if ((tree->gtFlags & GTF_SIDE_EFFECT) != 0)
+    {
+        if ((tree->gtFlags & GTF_SIDE_EFFECT) != GTF_EXCEPT)
+        {
+            return false;
+        }
+    }
+
+    // If relop's value is known, bail.
+    //
+    const ValueNum treeVN = tree->GetVN(VNK_Liberal);
+
+    if (vnStore->IsVNConstant(treeVN))
+    {
+        JITDUMP(" -- no, jump tree cond is constant\n");
+        return false;
+    }
+
+    // If there's just one statement, bail.
+    //
+    if (stmt == block->firstStmt())
+    {
+        JITDUMP(" -- no, no prior stmt\n");
+        return false;
+    }
+
+    JITDUMP("\noptRedundantRelop in " FMT_BB "; jump tree is\n", block->bbNum);
+    DISPTREE(jumpTree);
+
+    // We're going to search back to find the earliest tree in block that
+    //  * makes the current relop redundant;
+    //  * can safely and profitably forward substituted to the jump.
+    //
+    Statement* prevStmt      = stmt;
+    GenTree*   candidateTree = nullptr;
+    bool       reverse       = false;
+
+    // We need to keep track of which locals might be killed by
+    // the trees between the expression we want to forward substitute
+    // and the jump.
+    //
+    // We don't use a varset here because we are indexing by local ID,
+    // not by tracked index.
+    //
+    // The table size here also implicitly limits how far back we'll search.
+    //
+    enum
+    {
+        DEFINED_LOCALS_SIZE = 10
+    };
+    unsigned definedLocals[DEFINED_LOCALS_SIZE];
+    unsigned definedLocalsCount = 0;
+
+    while (true)
+    {
+        prevStmt = prevStmt->GetPrevStmt();
+
+        // Backwards statement walks wrap around, so if we get
+        // back to stmt we've seen everything there is to see.
+        //
+        if (prevStmt == stmt)
+        {
+            break;
+        }
+
+        // We are looking for ASG(lcl, ...)
+        //
+        GenTree* const prevTree = prevStmt->GetRootNode();
+
+        JITDUMP(" ... checking previous tree\n");
+        DISPTREE(prevTree);
+
+        // Ignore nops.
+        //
+        if (prevTree->OperIs(GT_NOP))
+        {
+            continue;
+        }
+
+        // If prevTree has side effects other than GTF_EXCEPT or GTF_ASG, bail.
+        //
+        if ((prevTree->gtFlags & GTF_SIDE_EFFECT) != (prevTree->gtFlags & (GTF_EXCEPT | GTF_ASG)))
+        {
+            JITDUMP(" -- prev tree has side effects\n");
+            break;
+        }
```

Comment:
It appears we only keep track of locals assigned at the root.

Do we need to check `(prevTreeRHS->gtFlags & GTF_ASG) == 0` to guard against nested definitions?